opengovsg/plumber
# PLU-125: [EXCEL-7] Create table row action

## Problem
Users want to insert a row of data into their excel file.

## Solution
See our M365 context doc on notion for more details, but essentially, we instruct users to use Excel [tables](https://support.microsoft.com/en-au/office/overview-of-excel-tables-7ab0bb7d-3a9e-4b56-a3c9-6c94334e492c) so that we can use the [createTableRow](https://learn.microsoft.com/en-us/graph/api/table-post-rows?view=graph-rest-1.0&tabs=http) API to insert rows into a spreadsheet.

This is the only data insertion API I've found that doesn't use cell coordinates; this works around the "read empty row coordinates " race without us having to make entire actions atomic.

That said, for race-free behaviour, we still need to use the same session ID for each excel file across our fleet. I've created a new redis DB called `APP_DATA` to store each file's session ID, and used a distributed redis lock to manipulate this safely. See the giant comment block in the code for more details.

## Tests
- Concurrently run multiple createTableRow calls on the same session ID (via k6) and check that there is no data loss.
- Check that the table row gets inserted successfully by the code, and data is inserted in the correct columns.
- Check that if new columns are inserted in the table, existing pipe still inserts data in the correct column (matching by column name).
- Check that an error is thrown if user tries to update the same column in different MultiRow entries.
- Check that Graph API errors are correctly parsed, and that sessions are re-created if we ever get a `invalidSessionReCreatable` error code

**Unit tests to be added in a later PR if team agrees with approach.**

## Deploy Notes
**New dependencies**:
- `redlock`: distributed lock
